packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,235 @@
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static addStackSetConstraint(scope: Resource, props: StackSetConstraintProps) {
+    const [pair, associationKey] = this.associationPrecheck(scope, props);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, props.description);
+    if (!(this.associationMap.get(associationKey).get(Constraints.LAUNCH_ROLE)) &&
+     !(this.associationMap.get(associationKey).get(Constraints.STACKSET))) {
+      const constraint = new CfnStackSetConstraint(scope, `StackSetConstraint${associationKey}`, {
+        acceptLanguage: props.acceptLanguage,
+        description: props.description ?? '',
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        accountList: props.accounts,
+        regionList: props.regions,
+        adminRole: props.adminRole.roleArn,
+        executionRole: props.adminRole.roleName,
+        stackInstanceControl: props.stackInstanceControlAllowed ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.STACKSET, constraint);
+    } else {
+      throw new Error(`Cannot have multiple launch or stackset constraints on association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: Resource, resourceId: string, tagOption: TagOption) {
+
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = getIdentifier(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const tO = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, tO);
+        }
+
+        new CfnTagOptionAssociation(scope, `TagOptionAssociation${getIdentifier(scope.node.addr, tagOptionKey)}`, {
+          resourceId: resourceId,
+          tagOptionId: this.tagOptionMap.get(tagOptionKey).ref,
+        });
+      });
+    });
+  }
+
+  private static associationMap = new Map();
+  private static tagOptionMap = new Map();
```

Comment:
Can you move the field declarations above the methods that use them? Otherwise, this code is pretty hard to read.

Also, why are these untyped?

Reply:
moved, added stronger types to them.